### PR TITLE
feat(ui5-popover,ui5-dialog): add accessibleRole property

### LINKED ISSUE #6090: [SF][Feature Request][Popover][Expected in ver.1.11.0]Enable changing the role of the Popover component
### **Feature Request Description**
Hi team
Now the Popover component has default `role=dialog` internally. Since the popover can also contain actionable/non-actionable elements, settled "role=dialog" to `Popover` seems not be the best practice, `Popover` may also contain a list of buttons and should have a "menu" role.(See discussion [here](https://github.com/SAP/ui5-webcomponents-react/issues/3793)). So can we have `accessibleRole` property of `Popover` component, to customize the role of `Popover` and its derived components like `ResponsivePopover` etc. ?

### **Proposed Solution**
Enable changing the accessibleRole of Popover component.

### **Additional Context**
Add other context or screenshots related to the feature request here.

### **Priority**
- [ ] Low
- [x] Medium
- [ ] High
- [ ] Very High

A clear and concise description of the impact/urgency of the required feature.

### **Stakeholder Info** (if applicable)

- Organization: SuccesssFactors
- Business impact: {...}

### PR DESCRIPTION
**Pull request by s-todorova**
Introduce a new property, which allows changing the role of the ui5-popover and ui5-dialog to either "dialog" (default), "alertdialog" or "none" (removes the role completely).

Fixes #6090 #5796

### CODE REVIEW

File: packages/main/src/Dialog.js [hunk 1]
```diff
@@ -343,6 +343,7 @@ class Dialog extends Popup {
 		this._isRTL = this.effectiveDir === "rtl";
 		this.onPhone = isPhone();
 		this.onDesktop = isDesktop();
+		this.accessibleRole = this._role;
```
Comment (gmkv) [inline packages/main/src/Dialog.js:346]: The attribute `accessible-role` cannot be customised with this approach.

**Review by gmkv — CHANGES_REQUESTED** [review]

**Comment by gmkv** [comment]
Now, any accessibleRole will be ignored if the 'state' property is Error or Warning. Let the dialog continue having its custom accessibleRole, if it has been set. If the property is not set, then keep the old behavior: 
- state=Error & state=Warning => `role='alertdialog'`
- state=anything else  => `role='dialog'`

File: packages/main/src/Popup.ts [hunk 2]
```diff
@@ -206,6 +206,16 @@ abstract class Popup extends UI5Element {
 	@property({ type: Boolean })
 	_blockLayerHidden!: boolean;
 
+	/**
+	 * Allows setting a custom role
+	 * @type {string}
+	 * @defaultvalue "dialog"
+	 * @public
+	 * @since 1.11.0
```
Comment (gmkv) [inline packages/main/src/Popup.ts:214]: 1.10.0

**Review by gmkv — CHANGES_REQUESTED** [review]

**Review by gmkv — APPROVED** [review]

**Comment by elenastoyanovaa** [comment]
Hello @SAP/ui5-webcomponents-topic-rd ,

Is this discussed with the accessibility expert and Stefan Schnabel, since this seems to not be present in the SR spec and in the UI5 Popover as well?

Elena

**Review by elenastoyanovaa — DISMISSED** [review]
Hello @SAP/ui5-webcomponents-topic-rd ,

Is this discussed with the accessibility expert and Stefan Schnabel, since this seems to not be present in the SR spec and in the UI5 Popover as well?

Elena

File: packages/main/src/Popup.ts [hunk 3]
```diff
@@ -188,6 +189,17 @@ abstract class Popup extends UI5Element {
 	@property({ defaultValue: "" })
 	accessibleNameRef!: string;
 
+	/**
+	 * Allows setting a custom role
```
Comment (gmkv) [inline packages/main/src/Popup.ts:193]: Update the comment to something like "Allows setting the role. Available options are:..." since this no longer allows to set an actual custom role.

File: packages/main/src/Dialog.ts [hunk 4]
```diff
@@ -340,7 +342,14 @@ class Dialog extends Popup {
 	}
 
 	get _role() {
-		return (this.state === ValueState.Error || this.state === ValueState.Warning) ? "alertdialog" : "dialog";
+		if (this.accessibleRole === PopupAccessibleRole.None) {
+			this.shadowRoot!.querySelector(".ui5-popup-root")?.removeAttribute("role");
+			this.shadowRoot!.querySelector(".ui5-popup-root")?.removeAttribute("aria-modal");
+		}
```
Comment (gmkv) [inline packages/main/src/Dialog.ts:348]: This getter does too much. Imagine accessing `const r = dialog._role;`and seeing the state being suddenly mutated.

File: packages/main/src/Popup.ts [hunk 5]
```diff
@@ -610,7 +622,11 @@ abstract class Popup extends UI5Element {
 	}
 
 	get _role() {
-		return "dialog";
+		if (this.accessibleRole === PopupAccessibleRole.None) {
+			this.shadowRoot!.querySelector(".ui5-popup-root")?.removeAttribute("role");
```
Comment (gmkv) [inline packages/main/src/Popup.ts:626]: Use the getter _root. But ideally you wouldn't need to call this at all.

File: packages/main/src/Popup.hbs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 <section
 	style="{{styles.root}}"
 	class="{{classes.root}}"
-	role="{{_role}}"
+	role={{_role}}
```
Comment (gmkv) [inline packages/main/src/Popup.hbs:4]: The role here ends up in written PascalCase.
https://www.deque.com/blog/wai-aria-top-6-mistakes-to-avoid/#:~:text=Authors%20need%20to%20remember%20that,in%20lowercase

**Review by gmkv — CHANGES_REQUESTED** [review]

File: packages/main/src/types/PopupAccessibleRole.ts [hunk 7]
```diff
@@ -14,21 +14,21 @@ enum PopupAccessibleRole {
...
+	dialog = "dialog",
 
 	/**
-	 * Represents the ARIA role "AlertDialog".
+	 * Represents the ARIA role "alertdialog".
 	 * @public
 	 * @type {AlertDialog}
 	 */
-	AlertDialog = "AlertDialog",
+	alertdialog = "alertdialog",
```
Comment (gmkv) [inline packages/main/src/types/PopupAccessibleRole.ts:31]: The enums should remain in PascalCase. Try to think of another way to make the role have all lowercase letters

**Review by gmkv — CHANGES_REQUESTED** [review]

File: packages/main/src/Dialog.ts [hunk 8]
```diff
@@ -289,7 +291,7 @@ class Dialog extends Popup {
 	}
 
 	get _ariaModal() {
-		return "true";
+		return (this.accessibleRole === PopupAccessibleRole.None) ? undefined : "true";
```
Comment (gmkv) [inline packages/main/src/Dialog.ts:294]: you don't need the parentheses here

File: packages/main/src/Dialog.ts [hunk 9]
```diff
@@ -340,7 +342,10 @@ class Dialog extends Popup {
 	}
 
 	get _role() {
-		return (this.state === ValueState.Error || this.state === ValueState.Warning) ? "alertdialog" : "dialog";
+		if (this.accessibleRole === PopupAccessibleRole.None) {
+			return undefined;
+		}
+		return (this.state === ValueState.Error || this.state === ValueState.Warning) ? PopupAccessibleRole.AlertDialog.toLowerCase() : this.accessibleRole.toLowerCase();
```
Comment (gmkv) [inline packages/main/src/Dialog.ts:348]: it would be easier to read if you write the state check in a separate if statement

File: packages/main/src/Popover.ts [hunk 10]
```diff
@@ -806,7 +807,7 @@ class Popover extends Popup {
 	}
 
 	get _ariaModal() {
-		return "true";
+		return (this.accessibleRole === PopupAccessibleRole.None) ? undefined : "true";
```
Comment (gmkv) [inline packages/main/src/Popover.ts:810]: same thing

File: packages/main/src/Popup.ts [hunk 11]
```diff
@@ -188,6 +189,18 @@ abstract class Popup extends UI5Element {
 	@property({ defaultValue: "" })
 	accessibleNameRef!: string;
 
+	/**
+	 * Allows setting a custom role. Available options are:
+	 * <code>Dialog</code>, <code>None</code> and <code>AlertDialog</code>.
```
Comment (gmkv) [inline packages/main/src/Popup.ts:194]: write those entries in a list
for example, see the docs here https://sap.github.io/ui5-webcomponents/playground/components/Avatar/

**Review by gmkv — CHANGES_REQUESTED** [review]
Please add short description for the PR

File: packages/main/src/Dialog.ts [hunk 12]
```diff
@@ -8,6 +8,7 @@ import {
 	isUpShift, isDownShift, isLeftShift, isRightShift,
 } from "@ui5/webcomponents-base/dist/Keys.js";
 import ValueState from "@ui5/webcomponents-base/dist/types/ValueState.js";
+import UI5Element from "@ui5/webcomponents-base/src/UI5Element.js";
```
Comment (gmkv) [inline packages/main/src/Dialog.ts:11]: this is unused

**Review by gmkv — APPROVED** [review]

File: packages/main/src/Popup.ts [hunk 13]
```diff
@@ -594,7 +611,7 @@ abstract class Popup extends UI5Element {
 	 * @abstract
 	 * @returns {string}
```
Comment (nnaydenow) [inline packages/main/src/Popup.ts:612]: returns { string | undefined } seems more correctly to me

File: packages/main/src/Dialog.ts [hunk 14]
```diff
@@ -339,8 +336,16 @@ class Dialog extends Popup {
...
 	}
 
-	get _role() {
-		return (this.state === ValueState.Error || this.state === ValueState.Warning) ? "alertdialog" : "dialog";
+	get _role(): string | undefined {
+		if (this.accessibleRole === PopupAccessibleRole.None) {
+			return undefined;
+		}
+
+		if (this.state === ValueState.Error || this.state === ValueState.Warning) {
```
Comment (elenastoyanovaa) [inline packages/main/src/Dialog.ts:344]: we should document this behavior - if a user sets a dialog role explicitly, we overwrite it. Is this the desired behavior? If yes, please add a note in the documentation of the property.

File: packages/main/test/specs/Popover.spec.js [hunk 15]
```diff
@@ -388,4 +388,14 @@ describe("Acc", () => {
 
 		assert.isNull(await popoverWithoutHeader.shadow$(".ui5-popup-root").getAttribute("aria-labelledby"), "Popover should NOT have aria-labelledby set.");
 	});
+
+	it("tests role setting from property", async () => {
+		const popover = await browser.$("#acc-role-popover");
+		await popover.removeAttribute("accessible-role");
+		assert.ok(await popover.shadow$("[role=dialog]"), "The default role is dialog.");
+
+		await popover.setAttribute("accessible-role", "menu");
```
Comment (elenastoyanovaa) [inline packages/main/test/specs/Popover.spec.js:397]: this should not be valid?

Comment (elenastoyanovaa) [reply packages/main/test/specs/Popover.spec.js:397]: we do not accept menu as a valid role

File: packages/main/test/specs/ResponsivePopover.mobile.spec.js [hunk 16]
```diff
@@ -18,3 +18,20 @@ describe("ResponsivePopover mobile general interaction", () => {
 		assert.ok(await popover.isDisplayedInViewport(), "Popover is opened.");
 	});
 });
+
+describe("Acc", () => {
```
Comment (elenastoyanovaa) [inline packages/main/test/specs/ResponsivePopover.mobile.spec.js:22]: those scenarios should not be valid.

File: packages/main/test/specs/ResponsivePopover.spec.js [hunk 17]
```diff
@@ -62,3 +62,18 @@ describe("ResponsivePopover general interaction", () => {
 		assert.notOk(await popover.isDisplayedInViewport(), "Popover is closed.");
 	});
 });
+
+describe("Acc", () => {
```
Comment (elenastoyanovaa) [inline packages/main/test/specs/ResponsivePopover.spec.js:66]: those scenarios should not be valid

**Review by elenastoyanovaa — CHANGES_REQUESTED** [review]
Please also add tests covering the alertdialog (with and without a set role as a property) and none scenario (we do not have tests if aria-modal is present or not)

**Review by TeodorTaushanov — APPROVED** [review]
